File: app/src/test/java/com/itis/android/lessontwo/ui/comics/ComicsPresenterTest.java
package com.itis.android.lessontwo.ui.comics;

import android.support.annotation.NonNull;

import com.itis.android.lessontwo.api.ApiFactory;
import com.itis.android.lessontwo.model.comics.Comics;
import com.itis.android.lessontwo.repository.ComicsRepository;
import com.itis.android.lessontwo.repository.ComicsRepositoryImpl;
import com.itis.android.lessontwo.repository.RepositoryProvider;
import com.itis.android.lessontwo.utils.RxUtils;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Single;

import static org.mockito.ArgumentMatchers.anyLong;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

/**
 * Created by a9 on 26.03.18.
 */

@RunWith(PowerMockRunner.class)
@PrepareForTest({RxUtils.class, ApiFactory.class})
public class ComicsPresenterTest {

    @SuppressWarnings("WeakerAccess")
    @Mock
    ComicsView$$State viewState;

    @SuppressWarnings("WeakerAccess")
    @Mock
    ComicsRepositoryImpl repository;

    private ComicsPresenter presenter;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        presenter = Mockito.spy(ComicsPresenter.class);
        presenter.setViewState(viewState);
        mockStatic(ApiFactory.class);
        mockStatic(RxUtils.class);
    }

    @Test
    public void onFirstViewAttach() throws Exception {
        // Arrange
        Mockito.doCallRealMethod().when(presenter).onFirstViewAttach();
        // Act
        presenter.onFirstViewAttach();
        // Assert
        Mockito.verify(viewState).getComicsId();
    }

    @Test
    public void loadComicsMockError() throws Exception {
        // Arrange
        RepositoryProvider.setComicsRepository(repository);
        Mockito.when(repository.comics(anyLong()))
                .thenReturn(Single.error(new Throwable()));
        // Act
        presenter.init(1);
        // Assert
        Mockito.verify(viewState).handleError(Mockito.any(Throwable.class));
    }

    @Test
    public void loadComicsMockSuccess() throws Exception {
        // Arrange
        RepositoryProvider.setComicsRepository(repository);
        Comics comics = Mockito.mock(Comics.class);
        Mockito.when(repository.comics(anyLong()))
                .thenReturn(Single.just(comics));
        // Act
        presenter.init(1);
        // Assert
        Mockito.verify(viewState).setImage(comics);
        Mockito.verify(viewState).setDescription(comics);
    }

    @Test
    public void loadComicsError() throws Exception {
        // Arrange
        List<Comics> comicsList = new ArrayList<>();
        Comics comics = new Comics();
        RepositoryProvider.setComicsRepository(new TestRepository(true, comicsList, comics));
        //Act
        presenter.init(1);
        // Assert
        Mockito.verify(viewState).handleError(Mockito.any(Throwable.class));
    }

    @Test
    public void loadComicsuccess() throws Exception {
        // Arrange
        List<Comics> comicsList = new ArrayList<>();
        Comics comics = new Comics();
        RepositoryProvider.setComicsRepository(new TestRepository(false, comicsList, comics));
        //Act
        presenter.init(1);
        // Assert
        Mockito.verify(viewState).setImage(comics);
        Mockito.verify(viewState).setDescription(comics);
    }

    @Test
    public void doActionInView() throws Exception {
        Mockito.verifyNoMoreInteractions(viewState);
    }

    @Test
    public void for100Coverage() {
        new ComicsPresenter$$ViewStateProvider().getViewState();
    }

    private class TestRepository implements ComicsRepository {

        private boolean error;
        private List<Comics> comicsList;
        private Comics comics;

        public TestRepository(boolean error, List<Comics> comicsList, Comics comics) {
            this.error = error;
            this.comicsList = comicsList;
            this.comics = comics;
        }

        @NonNull
        @Override
        public Single<List<Comics>> comics(Long offset, Long limit, String sort) {
            if (this.error) {
                return Single.error(new Throwable());
            } else {
                return Single.just(this.comicsList);
            }
        }

        @Override
        public Single<Comics> comics(Long id) {
            if (this.error) {
                return Single.error(new Throwable());
            } else {
                return Single.just(this.comics);
            }
        }
    }
}

File: app/src/test/java/com/itis/android/lessontwo/ui/comics/ComicsPresenterTest.kt
package com.itis.android.lessontwo.ui.comics

import com.itis.android.lessontwo.model.comics.Comics
import com.itis.android.lessontwo.repository.ComicsRepository
import io.mockk.*
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.reactivex.Single
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
class ComicsPresenterTest {

    @MockK
    lateinit var mockComicsRepository: ComicsRepository
    @MockK
    lateinit var mockViewState: `ComicsView$$State`

    lateinit var presenter: ComicsPresenter

    @BeforeEach
    fun setUp() {
        presenter = spyk(ComicsPresenter(mockComicsRepository))
        presenter.setViewState(mockViewState)
    }

    @Test
    @DisplayName("OOO MY GOSH IS A DOUBLE RAINBOW!!! I CAN BELIEVE IT")
    fun onFirstViewAttach() {
        val mockView = mockk<ComicsView>()
        every { mockViewState.getComicsId() } just Runs
        every { mockViewState.attachView(mockView) } just Runs

        presenter.attachView(mockView)

        verify {
            mockViewState.getComicsId()
        }
    }

    @Test
    @DisplayName("Шо пацаны ониме?!")
    fun whenInitExpectedSuccess() {
        // Arrange
        val expectedId = 5L
        val mockComics = mockk<Comics>()
        every { mockComicsRepository.comics(expectedId) } returns Single.just(mockComics)
        every { mockViewState.showProgress() } just Runs
        every { mockViewState.hideProgress() } just Runs
        every { mockViewState.setImage(mockComics) } just Runs
        every { mockViewState.setDescription(mockComics) } just Runs
        every { mockViewState.setPageCount(mockComics) } just Runs
        every { mockViewState.setPrice(mockComics) } just Runs
        // Act
        presenter.init(expectedId)
        // Assert
        verifyOrder {
            mockViewState.showProgress()
            mockViewState.hideProgress()
        }
        verify {
            mockViewState.setImage(mockComics)
            mockViewState.setDescription(mockComics)
            mockViewState.setPageCount(mockComics)
            mockViewState.setPrice(mockComics)
        }
    }

    @Test
    @DisplayName("should DIO muda muda muda")
    fun whenInitExpectedError() {
        // Arrange
        val expectedId = 566L
        val expectedError = mockk<Throwable>()
        val mockComics = mockk<Comics>()
        every { mockComicsRepository.comics(expectedId) } returns Single.error(expectedError)
        every { mockViewState.handleError(expectedError) } just Runs
        every { mockViewState.showProgress() } just Runs
        every { mockViewState.hideProgress() } just Runs
        // Act
        presenter.init(expectedId)
        // Assert
        verifyOrder {
            mockViewState.showProgress()
            mockViewState.hideProgress()
        }
        verify { mockViewState.handleError(expectedError) }
        verify(inverse = true) {
            mockViewState.setPrice(mockComics)
            mockViewState.setDescription(mockComics)
            mockViewState.setPageCount(mockComics)
            mockViewState.setImage(mockComics)
        }
    }
}
